lib/Sema/CSApply.cpp

```diff
@@ -7295,39 +7294,113 @@ static Expr *finishApplyDynamicCallable(ConstraintSystem &cs,
+    rewriter.finishArrayExpr(argumentTmp);
   } else {
+    // Get Key and Value associated types.
+    auto paramDictType = methodType->getParams()[0].getParameterType();
+    auto dictLitProto = cs.TC.getProtocol(
+        fn->getLoc(), KnownProtocolKind::ExpressibleByDictionaryLiteral);
+    auto conformance =
+        cs.TC.conformsToProtocol(paramDictType, dictLitProto, cs.DC,
+                                 ConformanceCheckFlags::InExpression);
+    auto keyAssocType =
+        ProtocolConformanceRef::getTypeWitnessByName(
+            paramDictType, *conformance, cs.getASTContext().Id_Key, &cs.TC)
+            ->getDesugaredType();
+    auto valueAssocType =
+        ProtocolConformanceRef::getTypeWitnessByName(
+            paramDictType, *conformance, cs.getASTContext().Id_Value, &cs.TC)
+            ->getDesugaredType();
+
     SmallVector<Identifier, 4> names;
     SmallVector<Expr *, 4> dictElements;
     for (unsigned i = 0, n = arg->getNumElements(); i < n; i++) {
-      Expr *labelExpr =
-        new (ctx) StringLiteralExpr(arg->getElementName(i).get(),
-                                    arg->getElementNameLoc(i),
-                                    /*Implicit*/ true);
-      Expr *pair =
-        TupleExpr::createImplicit(ctx, { labelExpr, arg->getElement(i) }, {});
+      auto *labelExpr = new (ctx) StringLiteralExpr(
+          arg->getElementName(i).get(), arg->getElementNameLoc(i),
+          /*Implicit*/ true);
+      cs.setType(labelExpr, keyAssocType);
+
+      Expr *pair = TupleExpr::createImplicit(
+          ctx,
+          {rewriter.visitStringLiteralExpr(labelExpr),
+           rewriter.coerceToType(
+               arg->getElement(i), valueAssocType,
+               loc.withPathElement(LocatorPathElt::getApplyArgToParam(0, 0)))},
+          {});
+
+      cs.setType(pair, TupleType::get({TupleTypeElt{keyAssocType},
+                                       TupleTypeElt{valueAssocType}},
+                                      ctx));
       dictElements.push_back(pair);
     }
-    argument = DictionaryExpr::create(ctx, SourceLoc(), dictElements, {},
-                                      SourceLoc());
+    auto *argumentTmp =
+        DictionaryExpr::create(ctx, SourceLoc(), dictElements, {}, SourceLoc());
+
+    cs.setType(argumentTmp, paramDictType);
+    rewriter.finishDictionaryExpr(argumentTmp);
+    argument = argumentTmp;
   }
   argument->setImplicit();
 
+  auto getType = [&](const Expr *E) -> Type { return cs.getType(E); };
+
   // Construct call to the `dynamicallyCall` method.
-  Expr *result = CallExpr::createImplicit(ctx, member, argument,
-                                          { argumentLabel });
-  cs.TC.typeCheckExpression(result, cs.DC);
-  cs.cacheExprTypes(result);
+  CallExpr *result = CallExpr::createImplicit(ctx, memberExpr, argument,
+                                              {argumentLabel}, getType);
+
+  cs.setType(result, methodType->getResult());
```

Comment:
I wonder if you've tried calling `cs.cacheExprTypes(result)` after `cs.setType(result, methodType->getResult())`?

`cs.cacheExprTypes` sets the types of the given expression and all subexpressions; this might eliminate the need for the code block below about "`tuple_expr` type not known in the constraint system".

Reply:
Thanks for the suggestion. That would indeed solve the problem. cacheExprTypes caches the whole tree into the constraint solver which is why I switched to this way. I see in other places that it passes both getType and setType into functions that parallel createImplicit. Something like that would be better.